hyperon_das/traverse_engines.py

```diff
@@ -0,0 +1,152 @@
+from abc import ABC, abstractmethod
+from random import choice
+from typing import TYPE_CHECKING, Any, Dict, List, Tuple, Union
+
+from hyperon_das.cache import (
+    ListIterator,
+    QueryAnswerIterator,
+    TraverselinksIterator,
+    TraverseNeighborsIterator,
+)
+from hyperon_das.exceptions import MultiplePathsError
+
+if TYPE_CHECKING:
+    from hyperon_das.das import DistributedAtomSpace
+
+
+class TraverseEngine(ABC):
+    def __init__(self, handle: str, **kwargs) -> None:
+        self.das: DistributedAtomSpace = kwargs['das']
+        self._cursor = handle
+
+    def _get_incoming_links(
+        self, **kwargs
+    ) -> List[Union[Tuple[Dict[str, Any], List[Dict[str, Any]]], Dict[str, Any]]]:
+        return self.das.get_incoming_links(atom_handle=self._cursor, **kwargs)
+
+    @abstractmethod
+    def get(self) -> Union[str, Dict[str, Any]]:
+        ...
+
+    @abstractmethod
+    def get_links(self, **kwargs) -> QueryAnswerIterator:
+        ...
+
+    @abstractmethod
+    def get_neighbors(self, **kwargs) -> QueryAnswerIterator:
+        ...
+
+    @abstractmethod
+    def follow_link(self, **kwargs) -> None:
+        ...
+
+    def goto(self, handle: str) -> None:
+        self._cursor = handle
+
+
+class HandleOnlyTraverseEngine(TraverseEngine):
+    def get(self) -> str:
+        return self._cursor
+
+    def get_links(self, **kwargs) -> QueryAnswerIterator:
+        incoming_links = self._get_incoming_links(handles_only=False, targets_document=True)
+        return TraverselinksIterator(
+            source=incoming_links, cursor=self._cursor, handles_only=True, **kwargs
+        )
+
+    def get_neighbors(self, **kwargs) -> List[str]:
+        filtered_targets_iterator = self.get_links(
+            link_type=kwargs.get('link_type'),
```

Comment:
kwargs should have defaults. Please fix all kwargs in all methods

Reply:
I already answered above